src/backend/metal/src/device.rs

```diff
@@ -1938,49 +1939,43 @@ impl hal::Device<Backend> for Device {
     fn destroy_image_view(&self, _view: n::ImageView) {
     }
 
-    // Emulated fence implementations
-    #[cfg(not(feature = "native_fence"))]
     fn create_fence(&self, signaled: bool) -> n::Fence {
-        Arc::new(n::FenceInner {
-            mutex: Mutex::new(signaled),
-            condvar: Condvar::new(),
-        })
+        n::Fence(RefCell::new(n::FenceInner::Idle { signaled }))
     }
     fn reset_fence(&self, fence: &n::Fence) {
-        *fence.mutex.lock() = false;
+        *fence.0.borrow_mut() = n::FenceInner::Idle { signaled: false };
     }
-    fn wait_for_fence(&self, fence: &n::Fence, timeout_ms: u32) -> bool {
+    fn wait_for_fence(&self, fence: &n::Fence, mut timeout_ms: u32) -> bool {
         debug!("wait_for_fence {:?} for {} ms", fence, timeout_ms);
-        let mut guard = fence.mutex.lock();
-        match timeout_ms {
-            0 => *guard,
-            0xFFFFFFFF => {
-                while !*guard {
-                    fence.condvar.wait(&mut guard);
-                }
-                true
+        let inner = fence.0.borrow();
+        let cmd_buf = match *inner {
+            native::FenceInner::Idle { signaled } => return signaled,
+            native::FenceInner::Pending(ref cmd_buf) => cmd_buf,
+        };
+        if timeout_ms == !0 {
```

Comment:
If I'm understanding correctly, I don't see how we can do these actions for a single fence in isolation, because we may be running multiple fences in parallel on the same thread (https://github.com/gfx-rs/gfx/blob/master/src/hal/src/device.rs#L528).

It seems like the `loop` would need to be around all running fences so we can start the timer from the same point. But even then with `WaitFor::Any` we want to race and avoid waiting for the longest completion or timeout, so we need to end the loop after any fence completes.

Reply:
Hmm, but this very call `wait_for_fence` is about a single fence, and the gluing of multiple fences lies on the shoulders of the default implementation. Are you saying that the very API is broken at the moment? I don't think the new implementation changing anything in this regard.

Reply:
👍 clarified on Gitter

As mentioned, we should reconsider the `sleep` policy and timing a bit more in the future, but it's not urgent at the moment.